partiql-parser/src/result.rs

```diff
@@ -2,33 +2,62 @@
 
 //! [`Error`] and [`Result`] types for parsing PartiQL.
 
-use std::fmt::Debug;
+use std::borrow::Cow;
+use std::fmt::{Debug, Display};
 
-use crate::lalr::Token;
-use crate::LexError;
 use partiql_source_map::location::Located;
 use thiserror::Error;
 
-/// General [`Result`] type for the PartiQL parser.
-pub type ParserResult<'input, T, Loc> = Result<T, ParserError<'input, Loc>>;
+/// Errors in the lexical structure of a PartiQL query.
+///
+/// ### Notes
+/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
+#[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum LexicalError<'input> {
+    /// Generic invalid input; likely an unrecognizable token.
+    #[error("Lexing error: invalid input `{}`", .0)]
+    InvalidInput(Cow<'input, str>),
+    /// Embedded Ion value is not properly terminated.
+    #[error("Lexing error: unterminated ion literal")]
+    UnterminatedIonLiteral,
+    /// Comment is not properly terminated.
+    #[error("Lexing error: unterminated comment")]
+    UnterminatedComment,
+    /// Any other lexing error.
+    #[error("Lexing error: unknown error")]
+    Unknown,
+}
 
-/// Errors from the PartiQL parser.
+/// Errors in the syntactic structure of a PartiQL query.
+///
+/// ### Notes
+/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
 #[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
+#[non_exhaustive]
 pub enum ParserError<'input, Loc>
 where
-    Loc: Debug, // TODO this should be `Loc: Display`
+    Loc: Display,
 {
     /// Indicates that there was a problem with syntax.
-    #[error("Syntax Error: {} at [{:?}]", _0.inner, _0.location)]
+    #[error("Syntax Error: {} at `{}`", _0.inner, _0.location)]
     SyntaxError(Located<String, Loc>),
 
+    /// There were not enough tokens to complete a parse
+    #[error("Unexpected end of input")]
+    UnexpectedEndOfInput,
```

Comment:
I saw the `Loc` argument was removed in the latest commit. Is there any value in having this argument for this error? I guess it will always point beyond the last parsed token.

PostgreSQL, for example, points out a syntax error when it expects another token before the end of input.
```
postgres=# SELECT * FROM foo WHERE;
ERROR:  syntax error at or near ";"
LINE 1: SELECT * FROM foo WHERE;
                               ^
postgres=#
```

Reply:
Dealing with the indices of the location outside of the source just got too tricky for the library and it's customers; I'd rather be explicit and say the input was exhausted unexpectedly.